### PR TITLE
search: adds functionality to remove search suggestion

### PR DESCRIPTION
**Pull request by comarquet**
Related to #497 

I modified parse_search_suggestions so that it retrieves the feedbackToken of each suggestion in the search bar and associates it with the number it is displayed when invoking get_search_suggestions, so that the function remove_search_suggestion can be called with the number as parameter to delete the corresponding entry. 

image

### CODE REVIEW

File: ytmusicapi/mixins/search.py [hunk 1]
```diff
@@ -7,6 +7,10 @@
 
 
 class SearchMixin(MixinProtocol):
+    def __init__(self):
+        self._latest_suggestions = None
+        self._latest_feedback_tokens = None
```
Comment (sigma67) [inline ytmusicapi/mixins/search.py:12]: Despite being an instance, we don't store request-specific state in the `YTMusic` instance between function calls.

The only state in YTMusic relates to authentication/configuration.

Please move the variables into function parameters.

File: ytmusicapi/mixins/search.py [hunk 2]
```diff
@@ -295,7 +299,8 @@ def get_search_suggestions(self, query: str, detailed_runs=False) -> Union[list[
                     {
                       "text": "d"
                     }
-                  ]
+                  ],
+                  "number": 1
```
Comment (sigma67) [inline ytmusicapi/mixins/search.py:303]: I'd prefer `index`, starting at 0

File: ytmusicapi/mixins/search.py [hunk 3]
```diff
@@ -329,6 +336,50 @@ def get_search_suggestions(self, query: str, detailed_runs=False) -> Union[list[
...
+        feedback_tokens: dict[int, str] = {}
+        search_suggestions = parse_search_suggestions(response, detailed_runs, feedback_tokens)
+
+        # Store the suggestions and feedback tokens for later use
+        self._latest_suggestions = search_suggestions
+        self._latest_feedback_tokens = feedback_tokens
 
         return search_suggestions
+
+    def remove_search_suggestion(self, number: int) -> bool:
```
Comment (sigma67) [inline ytmusicapi/mixins/search.py:349]: Same here, please rename `number` to `index`

File: ytmusicapi/mixins/search.py [hunk 4]
```diff
@@ -329,6 +336,50 @@ def get_search_suggestions(self, query: str, detailed_runs=False) -> Union[list[
...
+
+              # Assuming you want to remove suggestion number 1
+              success = ytmusic.remove_search_suggestion(number=1)
+              if success:
+                  print("Suggestion removed successfully")
+              else:
+                  print("Failed to remove suggestion")
+        """
+        if self._latest_suggestions is None or self._latest_feedback_tokens is None:
+            raise ValueError(
```
Comment (sigma67) [inline ytmusicapi/mixins/search.py:368]: Please raise `YTMusicError` or `YTMusicUserError` as you see fit

File: ytmusicapi/mixins/search.py [hunk 5]
```diff
@@ -329,6 +336,50 @@ def get_search_suggestions(self, query: str, detailed_runs=False) -> Union[list[
...
+
+        body = {"feedbackTokens": [feedback_token]}
+        endpoint = "feedback"
+
+        response = self._send_request(endpoint, body)
+
+        if "feedbackResponses" in response and response["feedbackResponses"][0].get("isProcessed", False):
+            return True
+
+        return False
```
Comment (sigma67) [inline ytmusicapi/mixins/search.py:385]:
```suggestion
        return bool(nav(response, ["feedbackResponses", 0, "isProcessed"], none_if_absent=True))
```

File: ytmusicapi/parsers/search.py [hunk 6]
```diff
@@ -257,17 +258,25 @@ def _get_param2(filter):
...
-def parse_search_suggestions(results, detailed_runs):
+def parse_search_suggestions(results, detailed_runs, feedback_tokens):
     if not results.get("contents", [{}])[0].get("searchSuggestionsSectionRenderer", {}).get("contents", []):
         return []
 
     raw_suggestions = results["contents"][0]["searchSuggestionsSectionRenderer"]["contents"]
     suggestions = []
 
+    count = 1  # Used for deleting a search suggestion
     for raw_suggestion in raw_suggestions:
```
Comment (sigma67) [inline ytmusicapi/parsers/search.py:269]:
```suggestion
    for index, raw_suggestion in enumerate(raw_suggestions):
```

idiomatic Python. Remove the count variable

File: ytmusicapi/parsers/search.py [hunk 7]
```diff
@@ -276,8 +285,14 @@ def parse_search_suggestions(results, detailed_runs):
...
+            suggestions.append({"text": text, "runs": runs, "fromHistory": from_history, "number": count})
         else:
             suggestions.append(text)
 
+        count += 1
+
     return suggestions
+
+
+def get_feedback_token(suggestion_number):
```
Comment (sigma67) [inline ytmusicapi/parsers/search.py:297]: not sure why this function exists

File: ytmusicapi/parsers/search.py [hunk 8]
```diff
@@ -4,6 +4,7 @@
 
 UNIQUE_RESULT_TYPES = ["artist", "playlist", "song", "video", "station", "profile", "podcast", "episode"]
 ALL_RESULT_TYPES = ["album", *UNIQUE_RESULT_TYPES]
+FEEDBACK_TOKENS: dict[int, str] = {}
```
Comment (sigma67) [inline ytmusicapi/parsers/search.py:7]: why is this here?

File: tests/mixins/test_search.py [hunk 9]
```diff
@@ -115,3 +115,27 @@ def test_search_library(self, config, yt_oauth):
             yt_oauth.search("beatles", filter="community_playlists", scope="library", limit=40)
         with pytest.raises(Exception):
             yt_oauth.search("beatles", filter="featured_playlists", scope="library", limit=40)
+
+    def test_remove_suggestion_valid(self, yt_auth):
+        first_pass = yt_auth.search("b")
+        assert len(first_pass) > 0, "Search returned no results"
```
Comment (sigma67) [inline tests/mixins/test_search.py:121]: I guess you're only calling `search` here to populate the history first? Could you add a comment stating this?

**Review by sigma67 — COMMENTED** [review]
Awesome, thanks for the contribution! And you even added tests :)

Looks pretty good, some minor comments and suggestions below.

**Comment by comarquet** [comment]
Hi ! Thank you for your feedback; it’s greatly appreciated. I’ve implemented all the changes you suggested.

**Comment by sigma67** [comment]
@comarquet thanks for your contribution! I've touched it up a bit and it's good to merge 🚀
